node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+}
+
+func (w *_node) LookupByString(key string) (ipld.Node, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeStruct:
+		field := typ.Field(key)
+		if field == nil {
+			return nil, ipld.ErrInvalidKey{
+				TypeName: typ.Name().String(),
+				Key:      basicnode.NewString(key),
+			}
+		}
+		fval := w.val.FieldByName(fieldNameFromSchema(key))
+		if !fval.IsValid() {
+			panic("TODO: go-schema mismatch")
+		}
+		if field.IsOptional() {
+			if fval.IsNil() {
+				return ipld.Absent, nil
+			}
+			fval = fval.Elem()
+		}
+		if field.IsNullable() {
+			if fval.IsNil() {
+				return ipld.Null, nil
+			}
+			fval = fval.Elem()
+		}
+		node := &_node{
+			schemaType: field.Type(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeMap:
+		var kval reflect.Value
+		switch ktyp := typ.KeyType().(type) {
+		case *schema.TypeString:
+			kval = reflect.ValueOf(key)
+		default:
+			asm := &_assembler{
+				schemaType: ktyp,
+				val:        reflect.New(w.val.Type().Key()).Elem(),
+			}
+			if err := (*_assemblerRepr)(asm).AssignString(key); err != nil {
+				return nil, err
+			}
+			kval = asm.val
+		}
+		fval := w.val.MapIndex(kval)
+		if !fval.IsValid() { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		if typ.ValueIsNullable() {
+			if fval.IsNil() {
+				return ipld.Null, nil
+			}
+			fval = fval.Elem()
+		}
+		node := &_node{
+			schemaType: typ.ValueType(),
+			val:        fval,
+		}
+		return node, nil
+	case *schema.TypeUnion:
+		var idx int
+		var mtyp schema.Type
+		for i, member := range typ.Members() {
+			if member.Name().String() == key {
+				idx = i
+				mtyp = member
+				break
+			}
+		}
+		if mtyp == nil { // not found
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		haveIdx := int(w.val.FieldByName("Index").Int())
+		if haveIdx != idx { // mismatching type
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		mval := w.val.FieldByName("Value").Elem()
+		node := &_node{
+			schemaType: mtyp,
+			val:        mval,
+		}
+		return node, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupByString",
+		// TODO
+	}
+}
+
+func (w *_node) LookupByNode(key ipld.Node) (ipld.Node, error) {
+	panic("TODO: LookupByNode")
+}
+
+func (w *_node) LookupByIndex(idx int64) (ipld.Node, error) {
+	switch typ := w.schemaType.(type) {
+	case *schema.TypeList:
+		if idx < 0 || int(idx) >= w.val.Len() {
+			return nil, ipld.ErrNotExists{
+				// TODO
+			}
+		}
+		// TODO: bounds check
+		val := w.val.Index(int(idx))
+		if typ.ValueIsNullable() {
+			if val.IsNil() {
+				return ipld.Null, nil
+			}
+			val = val.Elem()
+		}
+		return &_node{schemaType: typ.ValueType(), val: val}, nil
+	}
+	return nil, ipld.ErrWrongKind{
+		TypeName:   w.schemaType.Name().String(),
+		MethodName: "LookupByIndex",
+		// TODO
+	}
+}
+
+func (w *_node) LookupBySegment(seg ipld.PathSegment) (ipld.Node, error) {
+	panic("TODO: LookupBySegment")
```

Comment:
I imagined so, but I thought this PR was large enough as is :P 

Also, I wonder what would be the best way to test this, because the schema/gen/go tests don't use it. Maybe some of the selector test cases could also be made "generic" to run over both schema/gen/go and node/nodebind.

Reply:
Yep, those test batteries should still grow.  IIRC I did some work to make sure both iterators and lookup methods got exercised for the same data (with the newest form of tests, anyway), but maybe there's still a bit more to do there to make sure _all_ the lookup method variants are exercised too.

Getting selector tests to run over arbitrary nodes would also be super good.  I think there's a small barrier there right now, in that those tests only take one NodePrototype at the moment, and assume they can use it for both selector data and the test data, and that needs fixing... tractable but definitely enough of a diff to make it its own little quest.

Reply:
I'd be in favor to see this done in a subsequent PR once there are tests for that path.